### PR TITLE
[277] Post reading time

### LINKED ISSUE #277: [277] Post Reading Time
**Current state**
Reading times for post are unavailable on the homepage and show a question mark.

**Expected state**
Research has been done on post reading time algorithms. These reading times are now visible on the homepage for each post (instead of the question mark).

**TO DO**
- [ ] Create util for calculating and returning reading time
- [ ] Use util to calculate reading times in post cards

**Necessities**
- None

**Branch**
277-post-reading-time

### PR DESCRIPTION
**Pull request by MrGooss**
closes #277 

<br />

 **What is it supposed to do**  
Add a utility for calculating reading times. Should also display reading times for posts on the home screen.

<br />

 **How can we test your branch**  
Have a look at the home page, add a couple of posts of different sizes and check if the time is correct. WORDS_PER_MIN is set to 265 as suggested by Medium.

<br />

**1. General checklist**
- [x] a. Tested happy flow.
- [x] b. Tested unhappy flow.
- [x] c. No unexpected exceptions.
- [x] d. No code with high complexity. (Big O)
- [x] e. EOF newline.
- [x] f. No floating comma's and spaces.
- [x] g. No unexpected/unhandled errors.
- [x] h. Lines no longer than 'guideline' in IDE.
- [x] i. Correct indentation.
- [x] j. No usage of legacy functionalities.
- [x] k. (JS) Curly braces on the sameline as reserved keywords *(`if () {` instead of `{` on line below)*
- [x] l. (JS) No double quotes.
- [x] m. (JS) Single newline above unique code structures and reserved keywords.
- [x] n. (JS) Variables bundled where possible.
- [x] o. (JS) No semi-colons except where they have a functional necessity, such as for-loops.
- [x] p. (JS) No code that is meant exclusively for development, such as console.logs for data preview.

<br />

**2. If relevant, front-end checklist**
- [x] a. Responsive. (Desktop all the way to mobile 320px)
- [x] b. Using SCSS variables where relevant.
- [x] c. Relevant class names
- [x] d. No modified global SCSS properties, except if necessary and fully tested the impact.
- [x] e. Checked [caniuse](https://caniuse.com) for more modern CSS properties.

<br />

**3. If relevant, back-end checklist**
- [x] a. No unhandled promise rejections.
- [x] b. Seeders still functional.
- [x] c. No primary-key violations.
- [x] d. Normalization, atleast [3NF](https://en.wikipedia.org/wiki/Database_normalization).
- [x] e. Used guards (middleware) where necessary and tested if route is guarded.

<br />

**4. If relevant, test these browsers**
- [x] a. Chrome
- [x] b. Firefox
- [x] c. Opera
- [x] d. Edge
- [ ] e. Safari

---

**PR Rules**  
*1. Add relevant labels.*  
*2. Select atleast two reviewers.*  
*3. Assign all reviewers allowed to merge the branch, including yourself.*  
*4. If relevant to the sprint, put related issue in 'Awaiting Approval' column on project board.*  
*5. Add to the corresponding [milestone](https://github.com/AjobK/Seaqull/milestones).*

<br />

**Reviewer info**  
*1. You can use aliases to refer to specific PR checklist items. For example 'Hey @JohnDoe, for this line please check 3.a and 1.d.'*  
*2. Where possible, please use the suggestions feature in github so the author has a better idea of the solution you had in mind.*   

<br />

> This form should be filled in by the author of this PR.

### CODE REVIEW

**Comment by AjobK** [comment]
Waiting for new review request since things on homepage don't work too well.

**Comment by AjobK** [comment]
Please reopen when ready

**Comment by MrGooss** [comment]
PR is now ready, title will be changed

File: frontend/src/components/postsBlock/index.js [hunk 1]
```diff
@@ -20,17 +21,29 @@ class PostsBlock extends Component {
...
 		})
+
 		this.setState({
-			...this.state,
 			posts: convertedPosts
 		})
 	}
 
+	//TODO: add to util
+	getRawContentFromPostMetadata (metadata) {
```
Comment (AjobK) [inline frontend/src/components/postsBlock/index.js]:
```suggestion
	getRawContentFromPostMetadata(metadata) {
```

File: frontend/src/util/textUtil.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+    static getCalculatedWordReadTimeFromText(text, wordsPerMin = this.WORDS_PER_MIN) {
+        const wordCount = this.getWordCountFromText(text)
+        const wordTime = wordCount / wordsPerMin
+        
+        return wordTime
+    }
+
+    static getWordCountFromText(text) {
+        this.getTrimmedText(text)
```
Comment (AjobK) [inline frontend/src/util/textUtil.js]: Useless line?

**Review by AjobK — CHANGES_REQUESTED** [review]

File: frontend/src/components/postsBlock/index.js [hunk 3]
```diff
@@ -20,17 +21,29 @@ class PostsBlock extends Component {
...
 		})
+
 		this.setState({
-			...this.state,
 			posts: convertedPosts
 		})
 	}
 
+	//TODO: add to util
+	getRawContentFromPostData (data) {
```
Comment (AjobK) [inline frontend/src/components/postsBlock/index.js]:
```suggestion
	getRawContentFromPostData(data) {
```

File: frontend/src/components/postsBlock/index.js [hunk 4]
```diff
@@ -20,17 +21,29 @@ class PostsBlock extends Component {
 
 		this.state.posts.forEach((post) => {
 			try {
-				post.title = convertFromRaw(JSON.parse(post.title)).getPlainText()
+				post.title = this.getRawContentFromPostData(post.title)
+				post.readTime = TextUtil.getReadTimeFromText(this.getRawContentFromPostData(post.content))
```
Comment (AjobK) [inline frontend/src/components/postsBlock/index.js]:
```suggestion
				post.readTime = TextUtil.getReadTimeFromText(
				    this.getRawContentFromPostData(post.content)
				)
```

File: frontend/src/util/textUtil.js [hunk 5]
```diff
@@ -0,0 +1,40 @@
+class TextUtil {
+    // REF: 
+    // https://www.freecodecamp.org/news/how-to-more-accurately-estimate-read-time-for-medium-articles-in-javascript-fb563ff0282a/
+
+    static WORDS_PER_MIN = 265
+
+    static getReadTimeFromText(text) {
+        return this.getHumanizedReadingTimeFromText(this.getCalculatedWordReadTimeFromText(text))
```
Comment (AjobK) [inline frontend/src/util/textUtil.js]:
```suggestion
        return this.getHumanizedReadingTimeFromText(
            this.getCalculatedWordReadTimeFromText(text)
        )
```

**Review by AjobK — CHANGES_REQUESTED** [review]
Minor semantic things

File: frontend/src/components/postsBlock/index.js (hunk 1 above)
Comment (MrGooss) [reply frontend/src/components/postsBlock/index.js]: done

File: frontend/src/components/postsBlock/index.js (hunk 4 above)
Comment (MrGooss) [reply frontend/src/components/postsBlock/index.js]: Fixed

File: frontend/src/util/textUtil.js (hunk 5 above)
Comment (MrGooss) [reply frontend/src/util/textUtil.js]: Fixed

**Review by AjobK — APPROVED** [review]
Works well! Good job 😄

**Pull request merged**